jvm/ca.ualberta.maple.swan.drivers/src/resources/models.swirl

```diff
@@ -199,19 +211,505 @@ bb0(%0 : $`*Array<τ_0_0>`, %1 : $`Array<τ_0_0>`, %2 : $`@thin Array<τ_0_0>.Ty
+bb0(%0 : $`*τ_0_0`, %1 : $`*Set<τ_0_0>`):
+  set = pointer_read %1, $`Set<τ_0_0>`
+  elem = pointer_read set, $`τ_0_0`
+  pointer_write elem to %0
+  return elem
+}
+
+// %0 -> element that needs to be read, %1 -> index, %2 -> set
+// remove the element at index from set
+func [model] @`Swift.Set.remove(at: Swift.Set<A>.Index) -> A` : $`@out τ_0_0` {
+bb0(%0 : $`*τ_0_0`, %1 : $`@owned Set<τ_0_0>.Index`, %2 : $`*Set<τ_0_0>`):
+  set = pointer_read %2, $`Set<τ_0_0>`
+  elem = pointer_read set, $`τ_0_0`
+  pointer_write elem to %0
+  return elem
+}
+
+// return the index
+func [model] @`Swift.Set.startIndex.getter : Swift.Set<A>.Index` : $`@owned Set<τ_0_0>.Index` {
+bb0(%0 : $`Set<τ_0_0>`):
+  %ret = new $`@owned Set<τ_0_0>.Index`
+  return %ret
+}
+
+// %0 -> bool, %1 -> set that needs to be remove all the elements
+// remove all the elements from the set
+func [model] @`Swift.Set.removeAll(keepingCapacity: Swift.Bool) -> ()` : $`()` {
+bb0(%0 : $`Bool`, %1 : $`*Set<τ_0_0>`):
+  val = new $`τ_0_0`
+  set = pointer_read %1, $`Set<τ_0_0>`
+  pointer_write val to set
+  %ret = new $`()`
+  return %ret
+}
+
+// %0 -> set need to be combined, %1 -> set that combine other set
+// create a new set to combine two set
+func [model] @`Swift.Set.union<A where A == A1.Element, A1: Swift.Sequence>(__owned A1) -> Swift.Set<A>` : $`@owned Set<τ_0_0>` {
+bb0(%0 : $`*Set<τ_0_0>`, %1 : $`Set<τ_0_0>`):
+  new_set = new $`@owned Set<τ_0_0>`
+  data = pointer_read %1, $`τ_0_0`
+  pointer_write data to new_set
+  return new_set
+}
+
+// %0 -> array that contains elements for inserting, %1 -> set
+// insert the element into the set
+func [model] @`Swift.Set.formUnion<A where A == A1.Element, A1: Swift.Sequence>(__owned A1) -> ()` : $`()` {
+bb0(%0 : $`*Set<τ_0_0>`, %1 : $`*Set<τ_0_0>`):
+  set1 = pointer_read %0, $`Set<τ_0_0>`
+  elem = pointer_read set1, $`τ_0_0`
+  set2 = pointer_read %1, $`Set<τ_0_0>`
+  pointer_write [weak] elem to set2
+  %ret = new $`()`
+  return %ret
+}
+
+// %0 -> first set, %1 -> second set
+// get the intersection elements from two set and return a new set
+func [model] @`Swift.Set.intersection(Swift.Set<A>) -> Swift.Set<A>` : $`@owned Set<τ_0_0>` {
+bb0(%0 : $`@owned Set<τ_0_0>`, %1 : $`Set<τ_0_0>`):
+  new_set = new $`@owned Set<τ_0_0>`
+  data = pointer_read %1, $`τ_0_0`
+  pointer_write data to new_set
+  return new_set
+}
+
+// %0 -> array, %1 -> set
+// get the intersection elements from set and array, and return a new set
+func [model] @`Swift.Set.intersection<A where A == A1.Element, A1: Swift.Sequence>(A1) -> Swift.Set<A>` : $`@owned Set<τ_0_0>` {
+bb0(%0 : $`*Array<τ_0_0>`, %1 : $`Set<τ_0_0>`):
+  new_set = new $`@owned Set<τ_0_0>`
+  data = pointer_read %1, $`τ_0_0`
+  pointer_write data to new_set
+  return new_set
+}
+
+// %0 -> first set, %1 -> second set
+// directly change the set to intersected elements from two set
+func [model] @`Swift.Set.formIntersection<A where A == A1.Element, A1: Swift.Sequence>(A1) -> ()` : $`()` {
```

Comment:
This model doesn't look right. `elem` should not be of type `τ_0_0`, and `set` should not be of type `Set<τ_0_0>` (based on the given argument types).